EspressoSystems/HotShot
# [Upgradability] - Stop proposing and voting for successfully upgraded versions

Closes #3449.

### This PR: 
* Adds `decided_upgrade_certificate` to the upgrade task state, to track the globally shared version.
* Checks the version before proposing and voting, and only proposes and votes if the version upgrade isn't complete.

### Key places to review: 
* `crates/task-impls/src/upgrade.rs`.

### How to test this PR:
* Run with `--features example-upgrade` and see if:
  * Proposing is skipped for a decided upgrade.
  * The upgrade happens around view 25 for both dependency and non-dependency tasks.
* Upgrade task tests should also pass.
